csrc/id_model/id_model.h

```diff
@@ -121,11 +129,14 @@ class IdModel : public PolymorphicBase {
   // Returns if a self mapping was detected that would invalidate assumptions of
   // the overall lowering system.
   //
+  // It is assumed that for any tensor represented by a list of domains,
+  // those domains should never be mapped with each other. It may be
+  // possible to lift this assumption, but it's unclear if it could
+  // matter in practice.
+  //
   // TODO: Can we make this more of an alias analysis?
   // Ref: https://github.com/csarofeen/pytorch/pull/1954#discussion_r961940498
-  bool hasSelfMapping() const {
-    return self_mapping_info_.has_value();
-  }
+  std::optional<SelfMapping> hasSelfMapping(const TensorView& tv) const;
```

Comment:
nit: 
```suggestion
  std::optional<SelfMapping> hasSelfMapping(const TensorView* tv) const;
```

While I generally prefer reference types over pointer types whenever possible, the overall codebase pretty much always uses pointer types for historical reasons, and I think it's more important to keep it consistent.